src/main/java/org/spongepowered/api/command/parameter/managed/standard/VariableValueParameters.java

```diff
@@ -0,0 +1,389 @@
+         *
+         * @param choice The string to be matched
+         * @param returnedObject The {@link Object to return}
+         * @return This builder, for chaining
+         */
+        default StaticChoicesBuilder choice(String choice, Object returnedObject) {
+            return choices(Collections.singleton(choice), () -> returnedObject);
+        }
+
+        /**
+         * Adds a collection of choices to the parameter, along with the objects
+         * that would be returned if the associated choice is selected.
+         *
+         * @param choices A {@link Map} containing the choices and associated
+         *                objects.
+         * @return This builder, for chaining
+         */
+        default StaticChoicesBuilder choices(Map<String, ?> choices) {
+            for (Map.Entry<String, ?> entry : choices.entrySet()) {
+                choice(entry.getKey(), entry.getValue());
+            }
+
+            return this;
+        }
+
+        /**
+         * Adds a collection of choices to the parameter, along with a
+         * {@link Supplier} that will return an object if one of the choices is
+         * selected.
+         *
+         * @param choices The valid choices
+         * @param returnedObjectSupplier Supplies the object to return
+         * @return This builder, for chaining
+         */
+        StaticChoicesBuilder choices(Iterable<String> choices, Supplier<?> returnedObjectSupplier);
+
+        /**
+         * Sets what should happen if the usage of this parameter is requested.
+         *
+         * <p>If there are 5 or fewer choices available, and "showInUsage" is
+         * true, the choices will be shown in the command usage. Otherwise, the
+         * usage will only display only the key.</p>
+         *
+         * @param showInUsage true if the choices should be shown
+         * @return This builder, for chaining
+         */
+        StaticChoicesBuilder setShowInUsage(boolean showInUsage);
+
+        /**
+         * Tests for validity and creates this {@link ValueParameter}
+         *
+         * @return The {@link ValueParameter}
+         * @throws IllegalStateException if no choices have been specified
+         */
+        ValueParameter build();
+
+    }
+
+    /**
+     * A builder that creates a {@link ValueParameter} that tries to match an
+     * argument against a dynamic set of choices and returns an appropriate object
+     * based on the supplied argument.
+     */
+    public interface DynamicChoicesBuilder extends ResettableBuilder<ValueParameter, DynamicChoicesBuilder> {
+
+        /**
+         * Sets the parameter to get its choices from the supplied {@link Map},
+         * where each choice is associated with its own object.
+         *
+         * @param choices A supplier that returns an appropriate map.
+         * @return This builder, for chaining
+         */
+        DynamicChoicesBuilder setChoicesAndResults(Supplier<Map<String, ?>> choices);
```

Comment:
I'm following the suggestion by @Zidane [here](https://github.com/SpongePowered/SpongeAPI/pull/1587#discussion_r129314548) as allured to in my previous comment, so right now, I won't be changing it.